src/Front/Ast.hs

```diff
@@ -1,45 +1,31 @@
 module Front.Ast where
 
+import           Common.Identifiers             ( Identifiable(..)
+                                                , Identifier
+                                                )
 import           Common.Pretty
-import           Data.List                      ( intersperse )
-
--- | A type variable name (e.g., a, b)
-type TVarId = String
-
--- | A type constructor name (e.g., Int, Bool)
-type TConId = String
-
--- | A data constructor name (e.g., Just, False)
-type DConId = String
 
--- | A type class name (e.g., Eq, Ord)
-type TClassId = String
-
--- | A variable name (e.g., x, y, f)
-type VarId = String
-
--- | An operator name (e.g., +, foo)
-type OperatorId = String
+import           Data.List                      ( intersperse )
 
 -- | A complete program: a list of declarations
 newtype Program = Program [Definition]
   deriving (Eq, Show)
 
 -- | A value definition
 data Definition
-  = DefFn VarId [Pat] TypFn Expr
+  = DefFn Identifier [Pat] TypFn Expr
   | DefPat Pat Expr
   deriving (Eq, Show)
 
 -- | A pattern appearing on the LHS of a definition or match arm
 data Pat
-  = PatWildcard         -- ^ Match anything, i.e., @_@
-  | PatId VarId         -- ^ Bind value to variable, e.g., @v@
-  | PatLit Literal      -- ^ Literal match, e.g., @1@
-  | PatAs VarId Pat     -- ^ Pattern alias, e.g., @a \@ <pat>@
-  | PatTup [Pat]        -- ^ Match on a tuple, e.g., @(<pat>, <pat>)@
-  | PatCon DConId [Pat] -- ^ Match on data constructor, e.g., @Some <pat>@
```

Comment:
This is not something the parser even recognizes, so there's no point having this node. Instead, this is now expressed as a `PatApp` of `PatId`.